pkg/reconciler/pubsub/topic/topic.go

```diff
@@ -0,0 +1,441 @@
+
+func removeFinalizer(s *v1alpha1.Topic) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Delete(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *Reconciler) reconcilePublisher(ctx context.Context, topic *v1alpha1.Topic) (error, *servingv1.Service) {
+	name := resources.GeneratePublisherName(topic)
+	existing, err := r.serviceLister.Services(topic.Namespace).Get(name)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			logging.FromContext(ctx).Desugar().Error("Unable to get an existing publisher", zap.Error(err))
+			return err, nil
+		}
+		existing = nil
+	} else if !metav1.IsControlledBy(existing, topic) {
+		p, _ := json.Marshal(existing)
+		logging.FromContext(ctx).Desugar().Error("Got a pre-owned publisher", zap.Any("publisher", p))
+		return fmt.Errorf("Topic %q does not own service: %q", topic.Name, name), nil
+	}
+
+	tracingCfg, err := tracing.ConfigToJSON(r.tracingConfig)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Error serializing tracing config", zap.Error(err))
+	}
+
+	desired := resources.MakePublisher(&resources.PublisherArgs{
+		Image:         r.publisherImage,
+		Topic:         topic,
+		Labels:        resources.GetLabels(controllerAgentName, topic.Name),
+		TracingConfig: tracingCfg,
+	})
+
+	svc := existing
+	if existing == nil {
+		svc, err = r.ServingClientSet.ServingV1().Services(topic.Namespace).Create(desired)
+		if err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to create publisher", zap.Error(err))
+			return err, nil
+		}
+	} else if !equality.Semantic.DeepEqual(&existing.Spec, &desired.Spec) {
+		existing.Spec = desired.Spec
+		svc, err = r.ServingClientSet.ServingV1().Services(topic.Namespace).Update(existing)
+		if err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to update publisher", zap.Any("publisher", existing), zap.Error(err))
+			return err, nil
+		}
+	}
+	return nil, svc
+}
+
+func (r *Reconciler) getPublisher(ctx context.Context, topic *v1alpha1.Topic) (*serving.Service, error) {
+	pl, err := r.serviceLister.Services(topic.Namespace).List(resources.GetLabelSelector(controllerAgentName, topic.Name))
+
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Unable to list services: %v", zap.Error(err))
+		return nil, err
+	}
+	for _, pub := range pl {
+		if metav1.IsControlledBy(pub, topic) {
+			return pub, nil
+		}
+	}
+	return nil, apierrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+func (r *Reconciler) UpdateFromTracingConfigMap(cfg *corev1.ConfigMap) {
+	if cfg == nil {
+		r.Logger.Error("Tracing ConfigMap is nil")
+		return
+	}
+	delete(cfg.Data, "_example")
+
+	tracingCfg, err := tracingconfig.NewTracingConfigFromConfigMap(cfg)
+	if err != nil {
+		r.Logger.Warnw("failed to create tracing config from configmap", zap.String("cfg.Name", cfg.Name))
+		return
+	}
+	r.tracingConfig = tracingCfg
+	r.Logger.Debugw("Updated Tracing config", zap.Any("tracingCfg", r.tracingConfig))
+	// TODO: requeue all Topics.
```

Comment:
This can be tracked by the same issue as in the other requeue TODOs

Reply:
Done! added a link to the issue